File: app/java/src/main/java/com/example/production/domain/model/purchase/PurchaseOrder.java
package com.example.production.domain.model.purchase;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class PurchaseOrder {
    private Integer id;
    private String purchaseOrderNumber;
    private LocalDate orderDate;
    private String supplierCode;
    private String ordererCode;
    private String departmentCode;
    @Builder.Default
    private PurchaseOrderStatus status = PurchaseOrderStatus.CREATING;
    private String remarks;
    private LocalDateTime createdAt;
    private String createdBy;
    private LocalDateTime updatedAt;
    private String updatedBy;

    // リレーション
    private List<PurchaseOrderDetail> details;

    /**
     * 発注合計金額を取得する
     */
    public BigDecimal getTotalOrderAmount() {
        if (details == null || details.isEmpty()) {
            return BigDecimal.ZERO;
        }
        return details.stream()
                .map(d -> d.getOrderAmount() != null ? d.getOrderAmount() : BigDecimal.ZERO)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    /**
     * 最も早い入荷予定日を取得する
     */
    public LocalDate getEarliestDeliveryDate() {
        if (details == null || details.isEmpty()) {
            return null;
        }
        return details.stream()
                .map(PurchaseOrderDetail::getExpectedReceivingDate)
                .filter(d -> d != null)
                .min(Comparator.naturalOrder())
                .orElse(null);
    }
}

File: app/csharp/src/ProductionManagement.Domain/Models/Purchase/PurchaseOrder.cs
namespace ProductionManagement.Domain.Models.Purchase;

/// <summary>
/// 発注データ
/// </summary>
public class PurchaseOrder
{
    private string _statusValue = "作成中";

    public int Id { get; set; }
    public required string PurchaseOrderNumber { get; set; }
    public required DateOnly OrderDate { get; set; }
    public required string SupplierCode { get; set; }
    public string? OrdererCode { get; set; }
    public string? DepartmentCode { get; set; }

    /// <summary>
    /// DB から読み取った文字列値（Dapper 用）
    /// </summary>
    public string StatusValue
    {
        get => _statusValue;
        set => _statusValue = value;
    }

    /// <summary>
    /// ステータス（アプリケーション用）
    /// </summary>
    public PurchaseOrderStatus Status
    {
        get => PurchaseOrderStatusExtensions.FromDisplayName(_statusValue);
        set => _statusValue = value.GetDisplayName();
    }

    public string? Remarks { get; set; }
    public DateTime? CreatedAt { get; set; }
    public string? CreatedBy { get; set; }
    public DateTime? UpdatedAt { get; set; }
    public string? UpdatedBy { get; set; }

    // リレーション
    public IReadOnlyList<PurchaseOrderDetail> Details { get; set; } = [];
}
